akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,305 @@
+  case object ShuttingDown extends PoolState {
+    override def order: Int = 1
+  }
+
+  // does not accept new tasks, does not process tasks in queue
+  case object ShutDown extends PoolState {
+    override def order: Int = 2
+  }
+
+  // all threads have been stopped, does not process tasks and does not accept new ones
+  case object Terminated extends PoolState {
+    override def order: Int = 3
+  }
+}
+
+final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
+
+class AffinityPoolConfigurator(config: Config, prerequisites: DispatcherPrerequisites) extends ExecutorServiceConfigurator(config, prerequisites) {
+
+  sealed trait CPUAffinityStrategy {
+    def javaStrat: AffinityStrategies
+  }
+
+  case object NoAffinityStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.ANY
+  }
+  case object SameCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_CORE
+  }
+  case object SameSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_SOCKET
+  }
+  case object DifferentCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_CORE
+  }
+  case object DifferentSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_SOCKET
+  }
+
+  private def toAffinityStrategy(s: String): CPUAffinityStrategy = s match {
+    case "any"              ⇒ NoAffinityStrategy
+    case "same-core"        ⇒ SameCoreStrategy
+    case "same-socket"      ⇒ SameSocketStrategy
+    case "different-core"   ⇒ DifferentCoreStrategy
+    case "different-socket" ⇒ DifferentSocketStrategy
+    case x                  ⇒ throw new IllegalArgumentException("[%s] is not a valid cpu affinity strategy [any, same-core, same-socket, different-core, different-socket]!" format x)
+
+  }
+
+  private def toWaitingStrategy(s: String): WaitingStrategy = s match {
+    case "sleep"     ⇒ sleepWaitingStrategy
+    case "yield"     ⇒ yieldWaitingStrategy
+    case "busy-spin" ⇒ busySpinWaitingStrategy
+    case x           ⇒ throw new IllegalArgumentException("[%s] is not a valid waiting strategy[sleep, yield, same-socket, busy-spin]!" format x)
+
+  }
+
+  private val poolSize = ThreadPoolConfig.scaledPoolSize(
+    config.getInt("parallelism-min"),
+    config.getDouble("parallelism-factor"),
+    config.getInt("parallelism-max"))
+  private val affinityGroupSize = config.getInt("affinity-group-size")
+  private val strategies = config.getStringList("cpu-affinity-strategies").map(toAffinityStrategy)
+  private val waitingStrat = toWaitingStrategy(config.getString("worker-waiting-strategy"))
+  private val tf = new AffinityThreadFactory("affinity-thread-fact", strategies.map(_.javaStrat): _*)
```

Comment:
@zaharidichev I played with [OpenHFT Thread Affinity](https://github.com/OpenHFT/Java-Thread-Affinity) on our 48 cpu box (6 sockets, 8 CPUs each).

As I understand it our goal is to spread the (Java) threads over all available CPUs, and make them stick to the same CPU. I find the documentation difficult to understand but when using `SAME_CORE, SAME_SOCKET, SAME_CORE, SAME_SOCKET, ANY` I get the desired behavior up to 46 threads. CPU0 and CPU47 are "General use CPU", i.e. not available for affinity pinning.

Creating more threads than 46 will allocate them all to CPU46. One would expect some round-robin, but I guess it doesn't make sense to allocate more threads to this pool than number of available CPUs. We should enforce that.

I have used the following code to try this:
```scala
import net.openhft.affinity._
import net.openhft.affinity.AffinityStrategies._

class SleepRunnable(a2: AffinityLock) extends Runnable {
  def run() {
    a2.bind(false)
    try {
      (1 to 10).foreach { _ =>
        println(s"Thread ${Affinity.getThreadId} ${Thread.currentThread.getName} on cpu ${Affinity.getCpu}")
        Thread.sleep(5000);
      }
    } catch {
      case e: InterruptedException =>
        Thread.currentThread().interrupt();
    } finally {
        a2.release();
    }
  }
}


val a = AffinityLock.acquireLock();
try {
  (1 to 45).foreach { n =>
    // find a cpu on the same core, or the same socket, or any free cpu.
    val a2 = a.acquireLock(SAME_CORE, SAME_SOCKET, ANY);
    new Thread(new SleepRunnable(a2), s"T$n").start();
  } 
} finally {
  a.release()
}

Thread.sleep(1000)
println(AffinityLock.dumpLocks()
```

I will try the `AffinityThreadFactory` also.